### PR TITLE
Support query by block number

### LINKED ISSUE #415: Support query by block number (historical state)
**Problem:** 

Currently, subquery only supports storing the final state of index data. For example, we indexed a network called `ABC` and synchronized it to its latest height of 5000,  then the data in the database is reflected the result at this block height. 
But we don't know the state of the data when the altitude of 4000 is queried. Therefore, we need a way to record state changes with block heights. 

**Approach**

- Add an extra field to the table to record the range of blocks, eg [35, 38) means from block 35 to 37 it remain in this status. And when this record need to be "update", we create a new record with block range [38, ) 
- Insert a new record rather than update means we need to resolve the conflict of id, so rather let user created ID to primary key, we will generate another field to be PK. But we will face challenge for update entity relationships.  
- Ensure that the solution is scalable to current large projects (e.g. SubVis)

### PR DESCRIPTION
**Pull request by spitko**
Resolve #415 

- [x] Add `_id` and `_block_range` columns
- [x] Add exclude constraint on (`id`, `_block_range`)
- [x] Remove foreign keys and instead track relations in smart comments
- [x] Add sequelize hooks to keep `_block_range` up to date
- [x] Handle unique/other indexes
- [x] Fix subscription notification trigger

### CODE REVIEW

File: packages/node/src/indexer/store.service.ts [hunk 1]
```diff
@@ -80,6 +95,7 @@ export class StoreService {
   ): Promise<void> {
     this.schema = schema;
     this.modelsRelations = modelsRelations;
+    this.historical = argv['experimental-historical'];
```
Comment (stwiname) [inline packages/node/src/indexer/store.service.ts:98]: We need further checks that the project was started with this enabled, it should not be possible to add this in after a project has started indexing.

There also needs to be handling for if the project is started with this flag then it is no longer enabled. I think we can throw an error in this case.

File: packages/node/src/indexer/store.service.ts [hunk 2]
```diff
@@ -379,6 +505,28 @@ group by
     return rows.map((result) => camelCaseObjectKey(result)) as IndexField[];
   }
 
+  private async markAsDeleted(model: ModelStatic<any>, id: string) {
```
Comment (stwiname) [inline packages/node/src/indexer/store.service.ts:508]: This should use the transaction.

File: packages/node/src/indexer/store.service.ts [hunk 3]
```diff
@@ -249,6 +279,13 @@ export class StoreService {
           throw new Error('Relation type is not supported');
       }
     }
+    foreignKeyMap.forEach((keys, tableName) => {
```
Comment (stwiname) [inline packages/node/src/indexer/store.service.ts:282]: Should this be under a `this.historical` check?

**Review by stwiname — COMMENTED** [review]
Nice work @spitko, you've made good progress. 

I have left some comments, do you think it would be feasible to add some tests to the store service?

@jiqiang90 can you please have a look over this.

File: packages/node/src/indexer/store.service.ts (hunk 1 above)
Comment (spitko) [reply packages/node/src/indexer/store.service.ts]: done

File: packages/node/src/indexer/store.service.ts (hunk 2 above)
Comment (spitko) [reply packages/node/src/indexer/store.service.ts:547]: it does :+1:

File: packages/node/src/indexer/store.service.ts (hunk 3 above)
Comment (spitko) [reply packages/node/src/indexer/store.service.ts:290]: not really, it's not a no-op if historical=false

File: packages/node/src/indexer/store.service.ts [hunk 4]
```diff
@@ -249,17 +287,140 @@ export class StoreService {
...
+        .map((tags) => smartTags(tags, '|'))
+        .join('\n');
+      const query = commentTableQuery(`"${schema}"."${tableName}"`, comment);
+      extraQueries.push(query);
+    });
     if (this.config.proofOfIndex) {
       this.poiRepo = PoiFactory(this.sequelize, schema);
     }
     this.metaDataRepo = MetadataFactory(this.sequelize, schema);
+    this.checkHistoricalState();
```
Comment (jiqiang90) [inline packages/node/src/indexer/store.service.ts:301]: missing `await`

File: packages/node/src/indexer/store.service.ts [hunk 5]
```diff
@@ -249,17 +287,140 @@ export class StoreService {
...
 
     await this.sequelize.sync();
+    this.setMetadata('historicalStateEnabled', this.historical);
```
Comment (jiqiang90) [inline packages/node/src/indexer/store.service.ts:304]: missing `await`

**Comment by stwiname** [comment]
Hey @spitko 

Thank you for your work. We've tested this and we're happy with the functionality.

I'd like to request a couple of changes relating to the cli flags:

* On `query` can we please remove the `experimental-historical` flag and decide based on the db schemas whether this should be enabled and log whether it is enabled or not on startup
* On the `node`
    * Can we please default historical to be on
    * Remove `experimental-historical` and replace it with `disable-historical`, which would disable historical if provided
    * If the project was initially started without historical feature it should always be disabled
    * Log whether historical is enabled or not on startup

**Comment by spitko** [comment]
Made it so that the value of --disable-historical is only checked on first startup and the value from _metadata table (or false for pre-existing projects) is used afterwards.

**Review by stwiname — APPROVED** [review]

**Pull request merged**